### ISSUE TITLE
Parallel Write TFRecords Shards

### ISSUE THREAD

**Issue by rodyt** [tensorflow/tensorflow#36360]
**System information**
- TensorFlow version (you are using): TF2.0
- Are you willing to contribute it (Yes/No): No

I have a `tf.data.Dataset` that reads raw image and annotations files and shards them. However, using `tf.data.experimental.TFRecordWriter` writes sequentially and is currently very slow.

```
dataset = tf.data.Dataset.....
for i in range(num_shards):
    write_path = ....
    tfrecords_writer = tf.data.experimental.TFRecordWriter(
        filename=write_path,
        compression_type=compression_type
    )
    dataset_shard = dataset.shard(num_shards, shard_index)

    start_write_time = time.time()
```

I would like to be able to write the shards in parallel to reduce the write time by opening up multiple processes that process each shard individually.

I've tried using a `multiprocessing.Process` and starting a bunch of processes that way, but it doesn't seem to speed up writing.

**Comment by Leslie-Fang** [comment]
@rodyt Did you mean when you call tfrecords_writer.write it generate the tfrecord sequentially ?
Could you provide a piece of code and the running time as the baseline?

**Comment by Leslie-Fang** [comment]
Could you try to write each dataset shard into a different tfrecord file? I suspect it will be much faster.
Since it seems each tfrecords_writer.write op only have  one background thread which works. So even you start much process which invoke the same tfrecords_writer.write op, there may only  one thread works at each time.

**Comment by rodyt** [comment]
Hi @Leslie-Fang,

Thanks for your response and advice.

I am trying to write each shard into a different `tfrecord` file, as you mentioned. The `tf.data.Dataset` is reading from the file system, preprocessing, and sharding them at the end (it might be more efficient to put sharding at the beginning though).

I try to start **x** separate threads using `multiprocess` and then evenly distribute the shards across the threads. Each shard is written to its own `tfrecords` file, and each thread ends up writing **y** shards into **y** different `tfrecords` file (individually). (e.g. We have 4 shards and 2 processes; Process 0 will write shards 0 & 2, while process 1 writes shards 1 & 3)

This is the code I am running:

```
# Get the number of shards each process will be responsible for writing
num_shards_per_thread = num_shards // num_processes

coord = tf.train.Coordinator()
processes = []

# Create the processes
for thread_index in range(num_processes):
    params = (
        dataset,  # tf.data.Dataset object
        output_dir,  # str containing path to somewhere on filesystem
        num_shards,  # The total number of shards
        num_shards_per_thread,  # The number of shards each thread is responsible for
        thread_index
    )

    thread_process = multiprocessing.Process(
        target=save_shard,
        args=params
    )

    thread_process.start()
    processes.append(thread_process)
coord.join(processes)

```

Inside `save_shard`, I execute the following code:

```
# This code is executed for each process
for shard in range(num_shards_per_thread):

    # Get the current shard's index out of all shards that will be written
    # Takes into account the shard number in the current thread
    shard_index = shard + (num_shards_per_thread * thread_index)
    # Write each shard to its own .tfrecords file
    filename = "my_tfrecord_file_%d.tfrecords" % (shard_index)
    write_path = os.path.join(output_dir, filename)
    # Create the writer
    tfrecords_writer = tf.data.experimental.TFRecordWriter(
        filename=write_path
    )

    # Actually shard the dataset and pass in the calculated index from above
    dataset_shard = dataset.shard(num_total_shards, shard_index)

    # Do the actual writing
    tfrecords_writer.write(dataset_shard)

```

I tried to borrow some code from: [this SO post](https://stackoverflow.com/questions/51504222/writing-tfrecord-with-multithreading-is-not-fast-as-expected)

It doesn't work as expected. It looks like it hangs there and does nothing. It appears that the `tfrecord` files are created, albeit empty. I left the code running for a while and after I came back, nothing had been written yet.

I'm a bit of a newbie at multi-threading, so I could be trying to do it completely wrong.

**Comment by jsimsa** [comment]
The tf.data team has a plan for making it possible to read and write sharded TFRecord files in parallel. The first step in that direction is the tf.data snapshot proposal https://github.com/tensorflow/community/pull/193/files, which does this under the hoods when generating the snapshot. This is expected to be available in TF 2.3. As a follow up, we plan to expose API to `save` and `load` the snapshot (supporting TFRecord as the serialization format).

As a side note, my suggestion would be not to mix `multiprocessing.Process` with tf.data code. tf.data code is multi-threaded and mixing multi-processing with multi-threading is not safe (because only the calling thread in a forked process is copied to the new process).

**Comment by stomachacheGE** [comment]
I have the same problem, and there seems to be no cure at this moment.

**Comment by jsimsa** [comment]
Have you tried using the `tf.data.experimental.save` API that allows you to write the contents of a dataset in parallel (and later load it using `tf.data.experimental.load`)?

**Comment by iprovalo** [comment]
@jsimsa I am trying to understand how to make the` tf.data.experimental.save` work in parallel.  Do you have an example of this?  Thank you!

**Comment by jsimsa** [comment]
The implementation of `save` does parallel I/O. You as a user do not need to do anything.

**Comment by iprovalo** [comment]
@jsimsa thank you for your prompt answer!  I am trying to understand why I always get a single shard when I do the `save` operation.  If I try applying a custom shard function, like `shard_func=lambda x: x % 7` I get an error: `TypeError: <lambda>() takes 1 positional argument but 3 were given`. My tensor element structure is something like a tuple of tensors, a tensor, a tensor: 
```
(
(
TensorSpec(shape=(None, 1), dtype=tf.float32, name=None), TensorSpec(shape=(None, 2), dtype=tf.float32, name=None), TensorSpec(shape=(None, 2), dtype=tf.float32, name=None), TensorSpec(shape=(None, 10), dtype=tf.float32, name=None), TensorSpec(shape=(None, 20), dtype=tf.float32, name=None), TensorSpec(shape=(None, 4), dtype=tf.float32, name=None), TensorSpec(shape=(None, 1), dtype=tf.float32, name=None), TensorSpec(shape=(None, 2), dtype=tf.float32, name=None), TensorSpec(shape=(None, 16), dtype=tf.float32, name=None), TensorSpec(shape=(None, 1), dtype=tf.float32, name=None), TensorSpec(shape=(None, 1), dtype=tf.float32, name=None), TensorSpec(shape=(None, 1), dtype=tf.float32, name=None), TensorSpec(shape=(None, 1), dtype=tf.float32, name=None), TensorSpec(shape=(None, 1), dtype=tf.float32, name=None)
)
TensorSpec(shape=(None, 1), dtype=tf.float32, name=None), 
TensorSpec(shape=(None, 1), dtype=tf.float32, name=None)
)

```

**Comment by jsimsa** [comment]
If only one shard is being generated when you do not explicitly specify `shard_func`, it means that the TensorFlow runtime believes that there is only one CPU core on the machine that you are running on (as the number of CPU cores is used as the default number of shards [source](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/data/experimental/io_ops.cc#L145-L148)).

The error you are seeing is because your `shard_func` needs to have a signature that matches the structure of the elements flowing through the input pipeline. So you will need the following:

```
dataset = ... # dataset definition

def shard_func(x, y, z):
  shard_index = ... # computes index based on `x`, `y,` and `z`

tf.data.experimental.save(..., shard_func)
```

**Comment by iprovalo** [comment]
@jsimsa thank you!  I have tried this, I am getting: `AttributeError: 'Tensor' object has no attribute 'numpy'` (TF 2.4.0)

```
def shard_func(x, y, w, num_shards=10):
    shard_index = w[0].numpy() % num_shards
    return shard_index
```

`w[0]` because it is a batch.

**Comment by iprovalo** [comment]
a fix for this error: `tf.dtypes.cast(w[0], tf.int64) % num_shards`

**Comment by jsimsa** [comment]
For future reference, the `shard_func` argument will be traced and executed as graph computation (as per documentation). In other words, `w` will be a Tensor object, which does not have a `numpy()` method.

**Comment by iprovalo** [comment]
@jsimsa thank you!

I am still having some trouble with the save/load and shards.  I was able to use the shard function by using the features scores.  The shards came out very uneven.  

At the same time, I get this message on the GPU (not on the local integration test running on the CPU):

`2021-01-21 16:46:58.641448: W tensorflow/core/grappler/optimizers/data/auto_shard.cc:656] In AUTO-mode, and switching to DATA-based sharding, instead of FILE-based sharding as we cannot find appropriate reader dataset op(s) to shard. Error: Did not find a shardable source, walked to a node which is not a dataset: name: "LoadDataset/_1"
`

Is there a reference example somewhere where the save/load and sharding techniques are combined for the distributed dataset use case?

Thank you!

**Comment by iprovalo** [comment]
I was able to solve my initial uneven sharding issue with this (from the snapshot documentation):

```
dataset = dataset.enumerate() #this produces the id (x) for each record
tf.data.experimental.save(dataset, target_path, lambda x, y=x % num_shards)
```

Then, remove this `id` later:  `dataset = dataset.map(lambda x, y: y)`

I have a couple more related questions.  With these distributed sets, if I use repeat, it puts all the repeated data on disk.  It may be a large amount for many epochs and also takes much longer to generate.  Is there a way to feed this saved dataset to every epoch?  Before the distributed dataset, I was running no-repeat option and `steps_per_epoch = None`, which would use all the training data for every epoch.  

Also, for my unit-testing, is there an utils class I can use to create and save `SnapshotMetadataRecord`?

Thank you!

**Comment by iprovalo** [comment]
I think I got an answer to the first question:

`steps_per_epoch=int(round(train_size/batch_size/num_shards_train))`
`validation_steps=int(round(val_size/batch_size/num_shards_val))`

This takes care of running out of samples with the distributed dataset.  I think the number of steps has to be lesser than the dataset size (adjusted by the batch size and the number of shards).

**Comment by iprovalo** [comment]
@jsimsa I ended up writing a small [reference](https://github.com/iprovalo/tf/blob/main/input_pipeline_stats_benchmark.py) app to integrate all the parts for saving and loading the dataset as well as training a small model.  I am using it for benchmarking and profiling different approaches.  

I noticed a GPU underutilization for my app, it is unlikely related to the dataset pipeline, since it is completely preprocessed.  Here is [the issue - 46800](https://github.com/tensorflow/tensorflow/issues/46800).

**Comment by SmileTM** [comment]
```
import tensorflow as tf
import numpy as np


matrixs = np.random.randn(10000000, 512, 512, 3)

tfrecord_save_path = 'data.tfrecord'

with tf.io.TFRecordWriter(tfrecord_save_path) as tfwriter:
    for matrix in matrixs:
        example = tf.train.Example(features=tf.train.Features(
            feature={
                # 当需要存入矩阵时，有两种方法 一种是将矩阵Flatten 然后在读取的时候进行 形状reshape
                'matrix': tf.train.Feature(float_list=tf.train.FloatList(value=matrix.reshape(-1))),
                'matrix_shape': tf.train.Feature(int64_list=tf.train.Int64List(value=matrix.shape)),
            }
        ))

        tfwriter.write(example.SerializeToString())
```

Why doesn't TFRecordWriter have built-in integrated multiprocess write?  Maybe all in one `data.tfrecord` more  convenient, instead of multiple file "shards".   @jsimsa

**Comment by jsimsa** [comment]
@SmileTM please create a separate issue for this as `tf.io.TFRecordWriter` is orthogonal to tf.data.

For tf.data users, the recommended mechanism for efficiently serializing a tf.data.Dataset is to use `tf.data.experimental.save()` whose output can be later consumed using `tf.data.experimental.load()`.

**Comment by ofiryaish** [comment]
@jsimsa 

Since `tf.data.experimental.TFRecordWriter` IS DEPRECATED, tf.io.TFRecordWriter should be the only way to create TFRecords, and it does not scale well. Is there any solution for @SmileTM question?
